File: 056 Merge Intervals.cpp
#include <iostream>
#include <vector>
#include <algorithm>
/*
Merge Intervals
https://oj.leetcode.com/problems/merge-intervals/
*/
struct Interval {
    int start;
    int end;
};
Interval make_pair(int x,int y)
{
    Interval r;
    r.start = x;
    r.end = y;
    return r;
}
void dump(std::vector<Interval>& input)
{
    for(std::vector<Interval>::iterator i = input.begin();i!=input.end();i++)
    {
        std::cout<<"("<<i->start<<","<<i->end<<") ";
    }
    printf("\n");
}

bool compareIntevals(Interval i,Interval j)
{
    if(i.start != j.start) return i.start < j.start;
    else return i.end < j.end;
}

std::vector<Interval> mergeIntevals(std::vector<Interval>& input)
{
    std::vector<Interval> ret;
    if(input.size() == 0) return ret;
    
    std::stable_sort(input.begin(),input.end(),compareIntevals);
    
    std::vector<Interval>::iterator prev = input.begin();
    for(std::vector<Interval>::iterator i = input.begin()+1;i!=input.end();i++)
    {
        if(prev->end < i->start){
            ret.push_back(*prev);
            prev = i;
        } else if(prev->end <= i->end){
            prev->end = i->end;
        }
    }
    ret.push_back(*prev);
    return ret;
}
    
int main(int argc, char** argv)
{
    std::vector<Interval> input;
    input.push_back(make_pair(1,3));
    input.push_back(make_pair(2,6));
    input.push_back(make_pair(8,10));
    input.push_back(make_pair(15,18));
    
    std::vector<Interval> out = mergeIntevals(input);
    
    dump(input);
    dump(out);
    return 0;
}

File: 056 Merge Intervals.py
'''
56. Merge Intervals
https://leetcode.com/problems/merge-intervals/

Given an array of intervals where intervals[i] = [starti, endi],
merge all overlapping intervals, and return an array of the non-overlapping intervals
that cover all the intervals in the input.

Example 1:

Input: intervals = [[1,3],[2,6],[8,10],[15,18]]
Output: [[1,6],[8,10],[15,18]]
Explanation: Since intervals [1,3] and [2,6] overlap, merge them into [1,6].

Example 2:

Input: intervals = [[1,4],[4,5]]
Output: [[1,5]]
Explanation: Intervals [1,4] and [4,5] are considered overlapping.

Constraints:
    1 <= intervals.length <= 104
    intervals[i].length == 2
    0 <= starti <= endi <= 104
'''
class Solution:
    '''
    Real TikTok interview question: 20240125
    '''
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if not intervals:
            return []
        intervals.sort()
        result = []
        prev = intervals[0]
        for i in range(1, len(intervals)):
            curr = intervals[i]
            if curr[0] > prev[1]:
                # If current interval's starting point is bigger than
                # previous's end point, they are not overlapped
                result.append(prev)
                prev = curr
            else:
                # Else, they are overlapped. To combine current and previous,
                # the starting point is previous's starting point,
                # the end point is max of these 2's end point
                prev[1] = max(prev[1], curr[1])
        result.append(prev)
        return result
